pkg/agent/multicast/mcast_route.go

```diff
@@ -72,6 +75,8 @@ type MRouteClient struct {
 	nodeConfig                *config.NodeConfig
 	multicastInterfaces       []string
 	inboundRouteCache         cache.Indexer
+	inboundRouteCacheMutex    sync.RWMutex
```

Comment:
If `inboundRouteCacheMutex` is a mutex to guard inboundRouteCache, I don't think it is needed since the implementation of cache.Indexer has a ThreadSafeStore and an associated KeyFunc

Reply:
I use `inboundRouteCacheMutex` to ensure `DelMrouteEntry` won't be called twice for the same mroute.

Reply:
But it is actually a mutex working on the cache, not a route entry?

Reply:
Updated.

I decided to remove this lock. There is a chance that two goroutines calling `deleteInboundMRoute` for the same inbound mroute entry concurrently, which causes one `c.socket.DelMrouteEntry` failed because mroute cache has already been deleted by another goroutine.

```
func (c *MRouteClient) deleteInboundMRoute(mRoute *inboundMulticastRouteEntry) error {
	err := c.socket.DelMrouteEntry(net.ParseIP(mRoute.src), net.ParseIP(mRoute.group), mRoute.vif)
	if err != nil {
		return err
	}
	c.inboundRouteCache.Delete(mRoute)
	return nil
}
```


 This should not a problem because after the failure of calling `c.socket.DelMrouteEntry`, one goroutine, which is spawned by workqueue, will trigger its retry mechanism and keep processing in the next round. Another goroutine which spawned by `go wait.NonSlidingUntil(c.updateMrouteStats, time.Minute, stopCh)` could delete remaining staled routes in next minute.

```
	mEntries, _ := c.inboundRouteCache.ByIndex(GroupNameIndexName, group.String())
	for _, route := range mEntries {
		entry := route.(*inboundMulticastRouteEntry)
		err := c.deleteInboundMRoute(entry)
		if err != nil {
			return err
		}
	}
```


```
for _, inboundRoute := range deletedInboundRoutes {
		klog.V(2).InfoS("Deleting staled inbound multicast route", "group", inboundRoute.group, "source", inboundRoute.src)
		err := c.deleteInboundMRoute(inboundRoute)
		if err != nil {
			klog.ErrorS(err, "Failed to delete inbound multicast route", "group", inboundRoute.group, "source", inboundRoute.src, "VIF", inboundRoute.vif)
			continue
		}
	}
```

Also, The lock itself is not find-grained enough.